File: matiasmasso/Mat.NET/DT2/DTODoc.vb
Imports System.Windows

Public Class DTODoc
    Property Dest As List(Of String)
    Property Fch As Date
    Property Num As String
    Property Obs As List(Of String)
    Property Itms As List(Of DTODocItm)
    Property Cur As DTOCur
    Property Lang As DTOLang
    Property Dto As Decimal
    Property PuntsQty As Integer
    Property PuntsTipus As Decimal
    Property PuntsBase As DTOAmt
    Property Dpp As Decimal

    Property Valorat As Boolean
    Property IvaBaseQuotas As List(Of DTOTaxBaseQuota)
    Property RecarrecEquivalencia As Boolean
    Property Coletillas As List(Of String)

    Property DisplayTotalLogistic As Boolean
    Property DisplayPunts As Boolean
    Property WriteTemplate As Boolean
    Property Incoterm As DTOProveidor.Incoterms
    Property BOC As Boolean 'Ensobradora: Begining of Collection
    Property EOC As Boolean 'Ensobradora: End of Collection
    Property SelFeed1 As Boolean 'Ensobradora: End of Collection
    Property Estilo As Estilos

    Property CustomLines As List(Of String)

    Property SideLabel As SideLabels

    Public Enum Estilos
        Comanda
        Albara
        Factura
        Proforma
    End Enum

    Public Enum SumConcepts
        SumaDeImportes
        BaseImponible
        SumaAnterior
        SumaParcial
        SumaySigue
        Total
    End Enum

    Public Enum SideLabels
        None
        Export
        FacturaRectificativa
    End Enum

    Public Enum FontStyles
        Regular
        Bold
        Italic
        Underline
    End Enum


    Public Sub New(oEstilo As DTODoc.Estilos, oLang As DTOLang, oCur As DTOCur, Optional ByVal BlWriteTemplate As Boolean = True)
        MyBase.New()
        _Estilo = oEstilo
        _Lang = oLang
        _Cur = oCur
        _WriteTemplate = BlWriteTemplate
        _Dest = New List(Of String)
        _Obs = New List(Of String)
        _Itms = New List(Of DTODocItm)
        _IvaBaseQuotas = New List(Of DTOTaxBaseQuota)
    End Sub

    Public Function BackColor() As SixLabors.ImageSharp.Color
        Dim retval As SixLabors.ImageSharp.Color
        Select Case Estilo
            Case Estilos.Comanda
                retval = Color.Yellow
            Case Estilos.Albara
                retval = Color.LightSalmon
            Case Estilos.Factura
                retval = Color.SkyBlue
            Case Estilos.Proforma
                retval = Color.LightGreen
        End Select
        Return retval
    End Function

    Public Function DtoColumnDisplay() As Boolean
        Dim retVal As Boolean = _Itms.Exists(Function(x) x.Dto <> 0)
        Return retVal
    End Function

    Public Function Descomptes() As Boolean
        Return (_Dto <> 0) 'Or _Dpp <> 0 Or _PuntsTipus <> 0)
    End Function

    Public Function Impostos() As Boolean
        Return _IvaBaseQuotas.Count > 0
    End Function
End Class

Public Class DTODocItm
    Property Ref As String
    Property Text As String
    Property Qty As Integer
    Property Preu As DTOAmt
    Property Dto As Decimal
    Property Punts As Decimal
    Property Boxes As Integer
    Property m3 As Decimal
    Property Kg As Integer
    Property Hyperlink As String

    Property FontStyle As DTODoc.FontStyles
    Property MinLinesBeforeEndPage As Integer
    Property LeftPadChars As Integer

    Public Sub New(Optional ByVal sText As String = "", Optional ByVal oFontStyle As DTODoc.FontStyles = Nothing, Optional ByVal IntQty As Integer = 0, Optional ByVal oPreu As DTOAmt = Nothing, Optional ByVal SngDto As Decimal = 0, Optional ByVal SngPunts As Decimal = 0, Optional ByVal IntLeftPadChars As Integer = 0, Optional ByVal IntMinLinesBeforeEndPage As Integer = 0, Optional ByVal sRef As String = "", Optional Hyperlink As String = "")
        MyBase.New()
        _Ref = sRef
        _Text = sText
        If oFontStyle = Nothing Then
            _FontStyle = DTODoc.FontStyles.Regular ' oFontStyle
        Else
            _FontStyle = oFontStyle
        End If
        _Qty = IntQty
        _Preu = oPreu
        _Dto = SngDto
        _Punts = SngPunts
        _LeftPadChars = IntLeftPadChars
        _MinLinesBeforeEndPage = IntMinLinesBeforeEndPage
    End Sub

    Public Function Import() As DTOAmt
        Dim retval As DTOAmt = DTOAmt.FromQtyPriceDto(_Qty, _Preu, _Dto)
        Return retval
    End Function


End Class

File: matiasmasso/Mat.NET/DTO/DTODoc.cs

using System;
using System.Collections.Generic;
using System.Drawing;

namespace DTO
{
    public class DTODoc
    {
        public List<string> dest { get; set; }
        public DateTime fch { get; set; }
        public string num { get; set; }
        public List<string> obs { get; set; }
        public List<DTODocItm> itms { get; set; }
        public DTOCur cur { get; set; }
        public DTOLang lang { get; set; }
        public decimal dto { get; set; }
        public int puntsQty { get; set; }
        public decimal puntsTipus { get; set; }
        public DTOAmt puntsBase { get; set; }
        public decimal dpp { get; set; }

        public bool valorat { get; set; }
        public List<DTOTaxBaseQuota> ivaBaseQuotas { get; set; }
        public bool recarrecEquivalencia { get; set; }
        public List<string> coletillas { get; set; }

        public bool displayTotalLogistic { get; set; }
        public bool displayPunts { get; set; }
        public bool writeTemplate { get; set; }

        public DTOInvoice.ExportCods ExportCod { get; set; }

        public DTOIncoterm incoterm { get; set; }
        public bool BOC { get; set; } // Ensobradora: Begining of Collection
        public bool EOC { get; set; } // Ensobradora: End of Collection
        public bool selFeed1 { get; set; } // Ensobradora: End of Collection
        public Estilos estilo { get; set; }

        public List<string> customLines { get; set; }

        public SideLabels sideLabel { get; set; }

        public enum Estilos
        {
            comanda,
            albara,
            factura,
            proforma
        }

        public enum SumConcepts
        {
            sumaDeImportes,
            baseImponible,
            sumaAnterior,
            sumaParcial,
            sumaySigue,
            total
        }

        public enum SideLabels
        {
            none,
            export,
            facturaRectificativa
        }

        public enum FontStyles
        {
            regular,
            bold,
            italic,
            underline
        }


        public DTODoc(DTODoc.Estilos oEstilo, DTOLang oLang, DTOCur oCur, bool BlWriteTemplate = true) : base()
        {
            estilo = oEstilo;
            lang = oLang;
            cur = oCur;
            writeTemplate = BlWriteTemplate;
            dest = new List<string>();
            obs = new List<string>();
            itms = new List<DTODocItm>();
            ivaBaseQuotas = new List<DTOTaxBaseQuota>();
        }

        public Color BackColor()
        {
            Color retval = Color.Black;
            switch (estilo)
            {
                case Estilos.comanda:
                    {
                        retval = Color.Yellow;
                        break;
                    }

                case Estilos.albara:
                    {
                        retval = Color.LightSalmon;
                        break;
                    }

                case Estilos.factura:
                    {
                        retval = Color.SkyBlue;
                        break;
                    }

                case Estilos.proforma:
                    {
                        retval = Color.LightGreen;
                        break;
                    }
            }
            return retval;
        }

        public bool DtoColumnDisplay()
        {
            bool retVal = itms.Exists(x => x.Dto != 0);
            return retVal;
        }

        public bool Descomptes()
        {
            return (dto != 0); // Or _Dpp <> 0 Or _PuntsTipus <> 0)
        }

        public bool Impostos()
        {
            return ivaBaseQuotas.Count > 0;
        }
    }

    public class DTODocItm
    {
        public string Ref { get; set; }
        public string Text { get; set; }
        public int Qty { get; set; }
        public DTOAmt Preu { get; set; }
        public decimal Dto { get; set; }
        public decimal Punts { get; set; }
        public int Boxes { get; set; }
        public decimal m3 { get; set; }
        public int Kg { get; set; }
        public string Hyperlink { get; set; }

        public DTODoc.FontStyles FontStyle { get; set; }
        public int MinLinesBeforeEndPage { get; set; }
        public int LeftPadChars { get; set; }

        public DTODocItm(string sText = "", DTODoc.FontStyles oFontStyle = default(DTODoc.FontStyles), int IntQty = 0, DTOAmt oPreu = null/* TODO Change to default(_) if this is not a reference type */, decimal SngDto = 0, decimal SngPunts = 0, int IntLeftPadChars = 0, int IntMinLinesBeforeEndPage = 0, string sRef = "", string Hyperlink = "") : base()
        {
            Ref = sRef;
            Text = sText;
            if (oFontStyle == default(int))
                FontStyle = DTODoc.FontStyles.regular; // oFontStyle
            else
                FontStyle = oFontStyle;
            Qty = IntQty;
            Preu = oPreu;
            Dto = SngDto;
            Punts = SngPunts;
            LeftPadChars = IntLeftPadChars;
            MinLinesBeforeEndPage = IntMinLinesBeforeEndPage;
        }

        public DTOAmt Import()
        {
            DTOAmt retval = DTOAmt.FromQtyPriceDto(Qty, Preu, Dto);
            return retval;
        }
    }
}
